### PR TITLE
Kuttl tests for Ceph + fixes

### PR DESCRIPTION
**Pull request by nolancon**
### Description of your changes

Add kuttl tests for running against a Ceph cluster - these are not part of CI but can easily be run manually.

Also fixed:
- Issue where health-check buckets were not being removed from backend after ProviderConfig deletion.
- Non-empty buckets could not be deleted, now delete all objects first (see https://docs.aws.amazon.com/AmazonS3/latest/userguide/delete-bucket.html)

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: e2e/tests/ceph/provider-ceph/install-provider-cfg.sh [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+
+: "${AWS_ACCESS_KEY_ID?= required}"
+: "${AWS_SECRET_ACCESS_KEY?= required}"
+: "${CEPH_ADDRESS? = required}"
```
Comment (mhmxs) [inline e2e/tests/ceph/provider-ceph/install-provider-cfg.sh:5]:
```suggestion
: "${CEPH_ADDRESS?= required}"
```

File: internal/controller/providerconfig/backendstore_controller.go [hunk 2]
```diff
@@ -61,6 +62,19 @@ func (r *BackendStoreReconciler) Reconcile(ctx context.Context, req ctrl.Request
...
+			// healthcheck-controller to avoid a race condition between
+			// deleting the bucket from the backend and deleting the backend from
+			// the store. Ideally this deletion would be handled when the bucket
+			// object is garbage collected, however the bucket controller's Delete
+			// method is not triggered by garbage collected buckets.
+			// TODO: Investigate this behaviour on garbage collection.
+			hcBucketName := req.Name + healthCheckSuffix
+			r.log.Info("Deleting health check bucket", "name", hcBucketName)
+			if err := s3internal.DeleteBucket(ctx, r.backendStore.GetBackend(req.Name), aws.String(hcBucketName)); err != nil {
+				return ctrl.Result{}, err
```
Comment (mhmxs) [inline internal/controller/providerconfig/backendstore_controller.go:76]: Does it make sense to log and/or wrap this error? What if backend couldn't delete the health check? Backend in the store would be available for other go routines? Does it make sense to keep it in the store but mark as inactive, so we still can delete the health check, but we don't use the backend for new buckets?

File: internal/s3/bucket.go [hunk 3]
```diff
@@ -53,3 +65,146 @@ func BucketToPutBucketOwnershipControlsInput(bucket *v1alpha1.Bucket) *s3.PutBuc
...
+
+		if err := g.Wait(); err != nil {
+			return errors.Wrap(err, errDeleteObject)
+		}
+
+		// If the bucket contains many objects, the ListObjectsV2() call
+		// might not return all of the objects in the first listing. Check to
+		// see whether the listing was truncated. If so, retrieve the next page
+		// of objects and delete them.
+		if objects.IsTruncated {
```
Comment (mhmxs) [inline internal/s3/bucket.go:123]:
```suggestion
		if !objects.IsTruncated {
			break
		}
		
		objectsInput.ContinuationToken = objects.ContinuationToken
```

File: internal/s3/bucket.go [hunk 4]
```diff
@@ -53,3 +65,146 @@ func BucketToPutBucketOwnershipControlsInput(bucket *v1alpha1.Bucket) *s3.PutBuc
...
+
+		if err := g.Wait(); err != nil {
+			return errors.Wrap(err, errDeleteObject)
+		}
+
+		// If the bucket contains many objects, the ListObjectVersionsV2() call
+		// might not return all of the objects in the first listing. Check to
+		// see whether the listing was truncated. If so, retrieve the next page
+		// of objects and delete them.
+		if objectVersions.IsTruncated {
```
Comment (mhmxs) [inline internal/s3/bucket.go:164]:
```suggestion
		if !objectVersions.IsTruncated {
			break
		}
		
		objVersionsInput.VersionIdMarker = objectVersions.NextVersionIdMarker
		objVersionsInput.KeyMarker = objectVersions.NextKeyMarker
```

File: internal/s3/bucket.go [hunk 5]
```diff
@@ -53,3 +65,146 @@ func BucketToPutBucketOwnershipControlsInput(bucket *v1alpha1.Bucket) *s3.PutBuc
...
+			return nil
+		}
+	}
+
+	return err
+}
+
+func BucketExists(ctx context.Context, s3Backend *s3.Client, bucketName string) (bool, error) {
+	_, err := s3Backend.HeadBucket(ctx, &s3.HeadBucketInput{Bucket: aws.String(bucketName)})
+	if err != nil {
```
Comment (mhmxs) [inline internal/s3/bucket.go:193]:
```suggestion
	if err != nil {
		return false, resource.Ignore(isNotFound, err)
	}
```
Does this the same?

**Review by mhmxs — APPROVED** [review]
LGTM

File: internal/controller/providerconfig/backendstore_controller.go (hunk 2 above)
Comment (nolancon) [reply internal/controller/providerconfig/backendstore_controller.go:76]: Going to merge this PR but tracking this comment here #23 to address in separate PR

**Pull request merged**